### PR TITLE
LMS wolfBoot support.

### PR DESCRIPTION
**Pull request by philljj**
# Description

Adds support for LMS/HSS signatures to wolfBoot.

LMS/HSS is a post-quantum stateful hash-based signature scheme.

# Documentation

Added new docs post-quantum readme:
```
docs/PQ.md
```

# Building

See procedure in `docs/PQ.md`.

# Config

Added new LMS sim example
```
config/examples/sim-lms.config
```
The `LMS_LEVELS`, `LMS_HEIGHT`, and `LMS_WINTERNITZ`, `IMAGE_SIGNATURE_SIZE`, and (optionally) `IMAGE_HEADER_SIZE` must be set here.

```
SIGN?=LMS
...
LMS_LEVELS=2
LMS_HEIGHT=5
LMS_WINTERNITZ=8
...
IMAGE_SIGNATURE_SIZE=2644
IMAGE_HEADER_SIZE?=5288
```

In LMS the signature size is a function of the parameters. Use the added helper script `tools/lms/lms_siglen` to calculate your signature length given your LMS parameters:
```
$./tools/lms/lms_siglen 
levels:      3
height:      5
winternitz:  8
#
total_len:   3992
```

# More Info

See these links for more info on LMS and wolfSSL/wolfCrypt:
- https://github.com/wolfSSL/wolfssl-examples/tree/master/pq/stateful_hash_sig
- https://www.wolfssl.com/documentation/manuals/wolfssl/appendix07.html#post-quantum-stateful-hash-based-signatures

### CODE REVIEW

**Comment by dgarske** [comment]
Please resolve conflicts.

File: config/examples/sim-lms.config [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#
+# The acceptable parameter values are those in RFC8554:
+#   levels = {1..8}
+#   height = {5, 10, 15, 20, 25}
+#   winternitz = {1, 2, 4, 8}
+#
+# The number of available signatures is:
+#   N = 2 ** (levels * height)
+#
+# LMS/HSS Signature sizes are directly proportional to the levels parm,
```
Comment (anhu) [inline config/examples/sim-lms.config:13]: Remove parm.

File: config/examples/sim-lms.config [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+LMS_HEIGHT=5
+LMS_WINTERNITZ=8
+WOLFBOOT_SMALL_STACK=0
+SPI_FLASH=0
+DEBUG=0
+DELTA_UPDATES=0
+IMAGE_SIGNATURE_SIZE=2644
+IMAGE_HEADER_SIZE?=5288
+
+# it should be multiple of system page size
```
Comment (anhu) [inline config/examples/sim-lms.config:40]: "it" what is "it"?

File: options.mk [hunk 3]
```diff
@@ -286,6 +286,53 @@ ifeq ($(SIGN),RSA4096)
   endif
 endif
 
+ifeq ($(SIGN),LMS)
+  # In LMS the signature size is a function of the LMS parameters.
```
Comment (anhu) [inline options.mk:290]: In -> For

File: tools/lms/lms_siglen [hunk 4]
```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+#
```
Comment (anhu) [inline tools/lms/lms_siglen:2]: Do we need GPL boiler plate here? I don't know the answer so I'm asking.

**Review by anhu — CHANGES_REQUESTED** [review]

File: tools/lms/lms_siglen (hunk 4 above)
Comment (philljj) [reply tools/lms/lms_siglen:2]: Not sure. The shell scripts in `tools/scripts/` don't have GPL, so was using that as ref. But `lms_siglen` should probably have `.sh` suffix like them.

File: tools/lms/lms_common.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                ferror(file));
+        return WC_LMS_RC_WRITE_FAIL;
+    }
+
+    err = fclose(file);
+    if (err) {
+        fprintf(stderr, "error: fclose returned %d\n", err);
+        return WC_LMS_RC_WRITE_FAIL;
+    }
+
```
[the remark this answers is not in the document]
Comment (philljj) [reply tools/lms/lms_common.h:67]: Agree but probably should close and check return first  before reading to increase likelihood the write has actually reached persistent storage.

File: tools/lms/lms_siglen (hunk 4 above)
Comment (philljj) [reply tools/lms/lms_siglen:2]: I originally experimented with putting the sig len calculation logic in makefile scripts but was too messy. And even if sig len is automatically set, the user will likely till need to manually config the header length for their platform. Hence I settled on this helper script.

**Comment by dgarske** [comment]
@philljj , please add a docs/PQ.md or docs/LMS.md with the details from the PR description. Please include a link to the hash-sigs repo.

File: config/examples/sim-lms.config (hunk 2 above)
Comment (philljj) [reply config/examples/sim-lms.config]: I updated it to be similar to the other sim*.config files.

File: options.mk (hunk 3 above)
Comment (philljj) [reply options.mk]: done

File: docs/PQ.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+## LMS/HSS
+
+LMS/HSS is a post-quantum stateful hash-based signature scheme (HBS). It
+is known for having small public and private keys, but larger signatures.
+The signature size is tunable via the different LMS parameters.
+
+Stateful HBS schemes are based on the security of their underlying hash
+functions and Merkle trees, which are not expected to be broken by the advent
+of cryptographically relevant quantum computers
```
Comment (anhu) [inline docs/PQ.md:14]: Missing period at end of paragraph.

**Review by anhu — COMMENTED** [review]
Just one little punctuation problem.

File: tools/keytools/Makefile [hunk 7]
```diff
@@ -16,6 +16,14 @@ CFLAGS  += -I. -DWOLFSSL_USER_SETTINGS -I$(WOLFDIR) -I$(WOLFBOOTDIR)/include -DW
 LDFLAGS =
 OBJDIR = ./
 
+ifeq ($(SIGN),LMS)
```
Comment (dgarske) [inline tools/keytools/Makefile:19]: The `make keytools` isn't working for me. I built `hss_lib.a` as documented.

```
 hash-sigs % make
/usr/bin/gcc -Wall -O3 -c hss_thread_pthread.c -o hss_thread_pthread.o
/usr/bin/ar rcs hss_lib_thread.a hss.o hss_alloc.o hss_aux.o hss_common.o hss_compute.o hss_generate.o hss_keygen.o hss_param.o hss_reserve.o hss_sign.o hss_sign_inc.o hss_thread_pthread.o hss_verify.o hss_verify_inc.o hss_derive.o hss_zeroize.o lm_common.o lm_ots_common.o lm_ots_sign.o lm_ots_verify.o lm_verify.o endian.o hash.o sha256.o
/usr/bin/ar rcs hss_verify.a hss_verify.o hss_verify_inc.o hss_common.o hss_thread_single.o hss_zeroize.o lm_common.o lm_ots_common.o lm_ots_verify.o lm_verify.o endian.o hash.o sha256.o
/usr/bin/gcc -Wall -O3 demo.c hss_lib_thread.a -lcrypto -lpthread -o demo
/usr/bin/gcc -Wall -O3 test_hss.c test_testvector.c test_stat.c test_keygen.c test_sign.c test_sign_inc.c test_load.c test_verify.c test_verify_inc.c test_keyload.c test_reserve.c test_thread.c test_h25.c hss_lib_thread.a -lcrypto -lpthread -o test_hss
```

```
wolfboot % ls -la lib/hash-sigs/*.a
-rw-r--r--  1 davidgarske  staff  79256 Aug 23 12:25 lib/hash-sigs/hss_lib.a

wolfboot % cp config/examples/sim-lms.config .config
wolfboot % make distclean
wolfboot % make keytools V=1
...
gcc -Wall -Wextra -Werror -I. -DWOLFSSL_USER_SETTINGS -I../../lib/wolfssl -I../../include -DWOLFBOOT_KEYTOOLS -DWOLFBOOT_SIGN_LMS -DWOLFSSL_HAVE_LMS -DHAVE_LIBLMS -I../../lib/hash-sigs/ -D"LMS_LEVELS=2" -D"LMS_HEIGHT=5" -D"LMS_WINTERNITZ=8" -O2 -DIMAGE_HEADER_SIZE=5288 -DDELTA_UPDATES -c -o keygen.o keygen.c
Building signing tool
Building keygen tool
gcc -o sign sign.o ./asn.o ./aes.o ./ecc.o ./coding.o ./chacha.o ./ed25519.o ./ed448.o ./fe_operations.o ./ge_operations.o ./fe_448.o ./ge_448.o ./hash.o ./logging.o ./memory.o ./random.o ./rsa.o ./sp_int.o ./sp_c32.o ./sp_c64.o ./sha3.o ./sha256.o ./sha512.o ./tfm.o ./wc_port.o ./wolfmath.o ./ext_lms.o ./delta.o  -L../../lib/hash-sigs/ -l:hss_lib.a
gcc -o keygen keygen.o ./asn.o ./aes.o ./ecc.o ./coding.o ./chacha.o ./ed25519.o ./ed448.o ./fe_operations.o ./ge_operations.o ./fe_448.o ./ge_448.o ./hash.o ./logging.o ./memory.o ./random.o ./rsa.o ./sp_int.o ./sp_c32.o ./sp_c64.o ./sha3.o ./sha256.o ./sha512.o ./tfm.o ./wc_port.o ./wolfmath.o ./ext_lms.o ./delta.o  -L../../lib/hash-sigs/ -l:hss_lib.a
ld: library not found for -l:hss_lib.a
ld: library not found for -l:hss_lib.a
clangclang: : errorerror: : linker command failed with exit code 1 (use -v to see invocation)linker command failed with exit code 1 (use -v to see invocation)

make[1]: *** [keygen] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [sign] Error 1
make: *** [keytools] Error 2
```

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — DISMISSED** [review]
Over to @danielinux

**Comment by danielinux** [comment]
Unfortunately libhss can only be linked on a POSIX-C system and won't work baremetal:

```
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-sbrkr.o): in function `_sbrk_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/sbrkr.c:51: undefined reference to `_sbrk'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-writer.o): in function `_write_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/writer.c:49: undefined reference to `_write'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-closer.o): in function `_close_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/closer.c:47: undefined reference to `_close'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-fstatr.o): in function `_fstat_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/fstatr.c:55: undefined reference to `_fstat'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-isattyr.o): in function `_isatty_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/isattyr.c:52: undefined reference to `_isatty'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-lseekr.o): in function `_lseek_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/lseekr.c:49: undefined reference to `_lseek'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-readr.o): in function `_read_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/readr.c:49: undefined reference to `_read'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-abort.o): in function `abort':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/stdlib/../../../../../../../../newlib/libc/stdlib/abort.c:59: undefined reference to `_exit'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-signalr.o): in function `_kill_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/signalr.c:53: undefined reference to `_kill'
/usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/bin/ld: /usr/lib/gcc/arm-none-eabi/12.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc.a(lib_a-signalr.o): in function `_getpid_r':
/build/newlib-afIbHz/newlib-3.3.0/build/arm-none-eabi/thumb/v7-m/nofp/newlib/libc/reent/../../../../../../../../newlib/libc/reent/signalr.c:83: undefined reference to `_getpid'
```

**Review by danielinux — CHANGES_REQUESTED** [review]
Cannot merge until we figure out how to link a cross-compiled hss_lib.a that does not carry POSIX-C, OS system calls and other dependencies.

**Comment by dgarske** [comment]
Also we should have a CI test added for it so the example config is built (including that hash-sigs lib).

**Comment by philljj** [comment]
Note: these changes require `LMS_VERIFY_ONLY` support from this PR:
https://github.com/wolfSSL/wolfssl/pull/6738

File: tools/config.mk [hunk 8]
```diff
@@ -30,6 +30,9 @@ ifeq ($(ARCH),)
   DISABLE_BACKUP?=0
   WOLFBOOT_VERSION?=0
   V?=0
+  LMS_LEVELS?=0
```
Comment (dgarske) [inline tools/config.mk:33]: These also need added to the list at the bottom

**Review by dgarske — CHANGES_REQUESTED** [review]

File: tools/config.mk (hunk 8 above)
Comment (philljj) [reply tools/config.mk:33]: done!

**Review by danielinux — CHANGES_REQUESTED** [review]
POSIX dependencies now solved. Successfully tested on embedded target. 

Once the wolfSSL PR is merged, please add non-regression tests + github automations as suggested by @dgarske, and discussed internally.

**Review by danielinux — APPROVED** [review]

**Comment by danielinux** [comment]
Happy to merge if you think your change requests have been addressed properly @anhu @dgarske

**Review by dgarske — APPROVED** [review]

**Review by anhu — APPROVED** [review]

**Comment by anhu** [comment]
I'm good!! Thank you!

**Pull request merged**